synapse/handlers/room_list.py

```diff
@@ -295,19 +311,28 @@ def _append_room_entry_to_chunk(self, room_id, num_joined_users, chunk, limit,
                                     search_filter, from_federation=False):
         """Generate the entry for a room in the public room list and append it
         to the `chunk` if it matches the search filter
+
+        Args:
+            room_id (str): The ID of the room.
+            num_joined_users (int): The number of joined users in the room.
+            chunk (list)
+            limit (int): Maximum amount of rooms to display. Function will
+                return if length of chunk is greater than limit + 1.
+            search_filter (dict)
+            from_federation (bool): Whether this request originated from a
+                federating server or a client. Used for room filtering.
         """
         if limit and len(chunk) > limit + 1:
             # We've already got enough, so lets just drop it.
             return
 
-        result = yield self.generate_room_entry(room_id,
-            num_joined_users)
+        result = yield self.generate_room_entry(room_id, num_joined_users)
 
         if from_federation and not self.config.allow_non_federated_in_public_rooms:
-            if result["m.federate"] == False:
+            if not result or result["m.federate"] is False:
```

Comment:
`not result["m.federate"]` is clearer

Reply:
The problem was that if "m.federate" wasn't in `result`, it would imply a federating room.

I've realized this logic doesn't cover that and could lead to a crash, and thus have replaced with:

```
        if from_federation:
            if "m.federate" in result and not result["m.federate"]:
            # This is a room that other servers cannot join. Do not show them
            # this room.
            return
```

Reply:
well, m.federate will always be present in this case afaict.

However, for the record, what you want to handle the case you mention is `if not result.get("m.federate", True)`.

Also, consider merging it with the other if: `if from_federation and result...`

Reply:
Yep, split it up as two `and`s was getting unclean. This is cleaner though, thank you.